flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTaskNetworkInput.java

```diff
@@ -73,21 +72,20 @@
 
     private final DeserializationDelegate<StreamElement> deserializationDelegate;
 
-    private final RecordDeserializer<DeserializationDelegate<StreamElement>>[] recordDeserializers;
+    private final Map<InputChannelInfo, RecordDeserializer<DeserializationDelegate<StreamElement>>>
+            recordDeserializers;
+    private final Map<InputChannelInfo, Integer> flattenedChannelIndices;
```

Comment:
The only reason why we need this, is because of `StatusWatermarkValve`? And theoretically we could migrate `StatusWatermarkValve` from raw indexes to `InputChannelInfo` and get rid of this conversion as well?

If so, that brings another question. Could there be a similar bugs in `StatusWatermarkValve` that originate from using raw `int` as channel index?

Reply:
> The only reason why we need this, is because of StatusWatermarkValve? And theoretically we could migrate StatusWatermarkValve from raw indexes to InputChannelInfo and get rid of this conversion as well?

Yes, this is needed for `StatusWatermarkValve` which could also be migrated to `InputChannelInfo`.
However, I couldn't find any issues there (or any other classes).